kernel/src/checkpoint/mod.rs

```diff
@@ -1,8 +1,406 @@
+///
+/// [`Table::checkpoint`]: [`crate::table::Table::checkpoint`]
+pub struct CheckpointWriter {
+    /// Reference to the snapshot (i.e. version) of the table being checkpointed
+    pub(crate) snapshot: Arc<Snapshot>,
+}
+
+impl CheckpointWriter {
+    /// Creates a new CheckpointWriter from a snapshot
+    pub(crate) fn new(snapshot: Arc<Snapshot>) -> Self {
+        Self { snapshot }
+    }
+
+    /// Retrieves the checkpoint data and path information.
+    ///
+    /// This method generates the filtered actions for the checkpoint and determines
+    /// the appropriate destination path.
+    ///
+    /// # Returns
+    /// [`CheckpointData`] containing the checkpoint path and data to write.
+    ///
+    /// # Warning
+    /// All data must be written to persistent storage before calling `CheckpointWriter::finalize()`.
+    // This method is the core of the checkpoint generation process. It:
+    // 1. Determines whether to write a V1 or V2 checkpoint based on the table's
+    //    `v2Checkpoints` feature support
+    // 2. Reads actions from the log segment using the checkpoint read schema
+    // 3. Filters and deduplicates actions for the checkpoint
+    // 4. Chains the checkpoint metadata action if writing a V2 spec checkpoint
+    //    (i.e., if `v2Checkpoints` feature is supported by table)
+    // 5. Generates the appropriate checkpoint path
+    pub fn checkpoint_data(&mut self, engine: &dyn Engine) -> DeltaResult<CheckpointData> {
+        let is_v2_checkpoints_supported = self
+            .snapshot
+            .table_configuration()
+            .is_v2_checkpoint_write_supported();
+
+        let actions = self.snapshot.log_segment().read_actions(
+            engine,
+            CHECKPOINT_ACTIONS_SCHEMA.clone(),
+            CHECKPOINT_ACTIONS_SCHEMA.clone(),
+            None,
+        )?;
+
+        // Create iterator over actions for checkpoint data
+        let checkpoint_data =
+            CheckpointLogReplayProcessor::new(self.deleted_file_retention_timestamp()?)
+                .process_actions_iter(actions);
+
+        let version = self.snapshot.version().try_into().map_err(|e| {
+            Error::CheckpointWrite(format!(
+                "Failed to convert checkpoint version from u64 {} to i64: {}",
+                self.snapshot.version(),
+                e
+            ))
+        })?;
+
+        // Chain the checkpoint metadata action if using V2 checkpoints
+        let chained = checkpoint_data.chain(
+            is_v2_checkpoints_supported
+                .then(|| self.create_checkpoint_metadata_batch(version, engine)),
+        );
+
+        let checkpoint_path = ParsedLogPath::new_classic_parquet_checkpoint(
+            self.snapshot.table_root(),
+            self.snapshot.version(),
+        )?;
+
+        // Wrap the data iterator to send counts to the CheckpointWriter when dropped
+        let wrapped_iterator = CheckpointDataIterator {
+            inner: Box::new(chained),
```

Comment:
tiny nit: might defer the `chain` call to here:
```suggestion
            inner: Box::new(checkpoint_data.chain(checkpoint_metadata)),
```
(aside: conceptually it seems nicer to put metadata first followed by sidecars... but in practice the Delta spec doesn't require it so it's not actually meaningful)